### ISSUE TITLE
Port to CoreCLR & Roslyn

### ISSUE THREAD

**Issue by enlight** [enlight/klawr#39]
[CoreCLR](https://github.com/dotnet/coreclr/tree/master/Documentation) now has a [hosting API for Unix](https://github.com/dotnet/coreclr/blob/1d3b60d2c8e9451dbdcefab9834c67f20b0f2323/src/dlls/mscoree/unixinterface.cpp). It is more limited than the COM-based hosting API available on Windows but should be sufficient for getting Klawr working on Mac/Linux. There's an [open issue](https://github.com/dotnet/coreclr/issues/1256) to expose the Unix hosting API on Windows, and once that work is done there'll no longer be a need to directly interface with the COM hosting API on Windows.

Klawr currently makes use of COM for bootstrapping the CLR and for exposing some C# classes to native code, the latter usage is due to convenience rather than necessity as it reduces the amount of glue code one must write by hand. If there was a code generator for exposing managed classes to native code the COM dependency could be removed entirely, I don't know if one already exists, but by leveraging Roslyn it should be possible to build one that generates a native class from a managed class and glues the two together.

Another issue to consider is the lack of AppDomain(s) in CoreCLR, in the Windows MS CLR the only way to unload an assembly is to destroy the AppDomain it was loaded into, so that's what Klawr currently does to hot-reload the GameScripts assembly while running inside UnrealEd. In theory in CoreCLR the same thing can be accomplished using [AssemblyLoadContext](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs), though it's unclear if one can create and destroy those load contexts at will. Of course there's also the security/isolation that AppDomain(s) provide, in CoreCLR this would have to be accomplished by spawning separate processes.

Another interesting project is [DNX](https://github.com/aspnet/dnx), perhaps not something Klawr can use directly, but may be useful as a reference for proper usage of the CoreCLR hosting APIs and Roslyn.

**Comment by enlight** [comment]
The CoreCLR Unix hosting API is now available on Windows (see https://github.com/dotnet/coreclr/commit/480227becb804cd511c23bbcaf4994b839130fd8).

**Comment by rogersachan** [comment]
Has there been any progress on this?

**Comment by enlight** [comment]
I haven't worked on this project in 8 months, so I haven't been following CoreCLR development.
